File: core-extra/arazzo-parser/src/main/java/org/evomaster/arazzo/deserializer/AnyExpressionDeserializer.java
package org.evomaster.arazzo.deserializer;

import com.fasterxml.jackson.core.JacksonException;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JsonDeserializer;
import com.fasterxml.jackson.databind.JsonNode;
import org.evomaster.arazzo.models.domain.AnyExpression;

import java.io.IOException;

/**
 * Custom Jackson deserializer for {@link AnyExpression}.
 * It resolves dynamic JSON payloads by mapping plain text to a {@link AnyExpression.Expression},
 * and any other complex structure to a generic JSON node.
 */
public class AnyExpressionDeserializer extends JsonDeserializer<AnyExpression> {

    @Override
    public AnyExpression deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) throws IOException, JacksonException {
        JsonNode node = jsonParser.getCodec().readTree(jsonParser);

        if (node.isTextual()) {
            return new AnyExpression.Expression(node.asText());
        }
        JsonNode any = jsonParser.getCodec().treeToValue(node, JsonNode.class);
        return new AnyExpression.Any(any);
    }

}

File: core/src/main/kotlin/org/evomaster/core/problem/rest/arazzo/deserializer/AnyExpressionDeserializer.kt
package org.evomaster.core.problem.rest.arazzo.deserializer

import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.JsonDeserializer
import com.fasterxml.jackson.databind.JsonNode
import org.evomaster.core.problem.rest.arazzo.models.AnyExpression
import org.evomaster.core.problem.rest.arazzo.parser.ExpressionParser

class AnyExpressionDeserializer : JsonDeserializer<AnyExpression>() {
    override fun deserialize(
        p0: JsonParser,
        p1: DeserializationContext
    ): AnyExpression {
        val node: JsonNode = p0.codec.readTree(p0)
        if (node.isTextual && node.asText().startsWith("$")) {
            return try {
                val parsedExpression = ExpressionParser.parse(node.asText())
                AnyExpression.Expression(parsedExpression)
            } catch (e: Exception) {
                throw IllegalArgumentException("Arazzo Parsing Error: Invalid ${node.asText()}", e)
            }
        }

        return AnyExpression.Constant(node)
    }
}